libobs/audio-monitoring/nix/pulseaudio-wrapper.c

```diff
@@ -0,0 +1,333 @@
+			pa_threaded_mainloop_free(pulseaudio_mainloop);
+			pulseaudio_mainloop = NULL;
+		}
+	}
+
+	pthread_mutex_unlock(&pulseaudio_mutex);
+}
+
+void pulseaudio_lock()
+{
+	pa_threaded_mainloop_lock(pulseaudio_mainloop);
+}
+
+void pulseaudio_unlock()
+{
+	pa_threaded_mainloop_unlock(pulseaudio_mainloop);
+}
+
+void pulseaudio_wait()
+{
+	pa_threaded_mainloop_wait(pulseaudio_mainloop);
+}
+
+void pulseaudio_signal(int wait_for_accept)
+{
+	pa_threaded_mainloop_signal(pulseaudio_mainloop, wait_for_accept);
+}
+
+void pulseaudio_accept()
+{
+	pa_threaded_mainloop_accept(pulseaudio_mainloop);
+}
+
+int_fast32_t
+pulseaudio_get_source_info_list(pa_source_info_cb_t cb, void *userdata)
+{
+	if (pulseaudio_context_ready() < 0)
+		return -1;
+
+	pulseaudio_lock();
+
+	pa_operation *op = pa_context_get_source_info_list(
+		pulseaudio_context, cb, userdata);
+	if (!op) {
+		pulseaudio_unlock();
+		return -1;
+	}
+	while (pa_operation_get_state(op) == PA_OPERATION_RUNNING)
+		pulseaudio_wait();
+	pa_operation_unref(op);
+
+	pulseaudio_unlock();
+
+	return 0;
+}
+
+int_fast32_t
+pulseaudio_get_source_info(pa_source_info_cb_t cb, const char *name,
+                           void *userdata)
+{
+	if (pulseaudio_context_ready() < 0)
+		return -1;
+
+	pulseaudio_lock();
+
+	pa_operation *op = pa_context_get_source_info_by_name(
+		pulseaudio_context, name, cb, userdata);
+	if (!op) {
+		pulseaudio_unlock();
+		return -1;
+	}
+	while (pa_operation_get_state(op) == PA_OPERATION_RUNNING)
+		pulseaudio_wait();
+	pa_operation_unref(op);
+
+	pulseaudio_unlock();
+
+	return 0;
+}
+
+int_fast32_t pulseaudio_get_server_info(pa_server_info_cb_t cb, void *userdata)
+{
+	if (pulseaudio_context_ready() < 0)
+		return -1;
+
+	pulseaudio_lock();
+
+	pa_operation *op = pa_context_get_server_info(
+		pulseaudio_context, cb, userdata);
+	if (!op) {
+		pulseaudio_unlock();
+		return -1;
+	}
+	while (pa_operation_get_state(op) == PA_OPERATION_RUNNING)
+		pulseaudio_wait();
+	pa_operation_unref(op);
+
+	pulseaudio_unlock();
+	return 0;
+}
+
+pa_stream *pulseaudio_stream_new(const char *name, const pa_sample_spec *ss,
+                                 const pa_channel_map *map)
+{
+	if (pulseaudio_context_ready() < 0)
+		return NULL;
+
+	pulseaudio_lock();
+
+	pa_proplist *p = pulseaudio_properties();
+	pa_stream *s = pa_stream_new_with_proplist(
+		pulseaudio_context, name, ss, map, p);
+	pa_proplist_free(p);
+
+	pulseaudio_unlock();
+	return s;
+}
+
+int_fast32_t pulseaudio_connect_playback(pa_stream *s, const char *name, const
+pa_buffer_attr *attr, pa_stream_flags_t flags)
+{
+	if (pulseaudio_context_ready() < 0)
+		return -1;
+
+	size_t dev_len = strlen(name) - 8;
+	char device[dev_len];
+	memcpy(device, name, dev_len);
+	device[dev_len] = '\0';
+
+	pulseaudio_lock();
+	int_fast32_t ret = pa_stream_connect_playback(s, device, attr, flags,
+	                                              NULL, NULL);
+	pulseaudio_unlock();
+	return ret;
+}
+
+void
```

Comment:
Style: Don't line break between the function return type and the function name in a function declaration.